In [3]:
import numpy as np
import pandas as pd

In [36]:
file_path = "~/task-2b-failures-in-cellular-apoptosis/data/train_data.data"
# above .data file is comma delimited
data = pd.read_csv(file_path, delimiter=",", low_memory=False)

In [37]:
data.head()

Unnamed: 0                     0                       1  \
0           0  -0.16099999845027924   -0.014000000432133675   
1           1  -0.15800000727176666  -0.0020000000949949026   
2           2                     ?                       ?   
3           3  -0.16899999976158142    -0.02500000037252903   
4           4  -0.18299999833106995   -0.050999999046325684   

                       2                      3                      4  \
0  0.0020000000949949026  -0.035999998450279236  -0.032999999821186066   
1  -0.012000000104308128   -0.02500000037252903  -0.012000000104308128   
2                      ?                      ?                      ?   
3  -0.009999999776482582   -0.04100000113248825   -0.04500000178813934   
4  -0.023000000044703484   -0.07699999958276749   -0.09200000017881393   

                       5                     6                     7  \
0   -0.09300000220537186   0.02500000037252903  0.004999999888241291   
1   -0.10599999874830246  0.013000000268220901  0.004999999888241291   
2                      ?                     ?                     ?   
3    -0.0689999982714653   0.03799999877810478  0.014000000432133675   
4  -0.014999999664723873   0.07100000232458115  0.027000000700354576   

                      8  ...                   5399                  5400  \
0                   0.0  ...   0.006000000052154064  0.013000000268220901   
1                   0.0  ...  0.0020000000949949026  -0.00800000037997961   
2                     ?  ...                      ?                     ?   
3   0.00800000037997961  ...    0.01899999938905239  0.009999999776482582   
4  0.019999999552965164  ...   0.050999999046325684  0.012000000104308128   

                   5401                  5402                  5403  \
0  0.020999999716877937  0.019999999552965164   0.01600000075995922   
1  0.007000000216066837  0.014999999664723873  -0.00800000037997961   
2                     ?                     ?                     ?   
3   0.02500000037252903   0.02500000037252903  0.020999999716877937   
4   0.05000000074505806   0.03799999877810478  0.050999999046325684   

                    5404                   5405                  5406  \
0  -0.010999999940395355   0.003000000026077032  0.009999999776482582   
1  -0.010999999940395355  -0.004000000189989805  0.013000000268220901   
2                      ?                      ?                     ?   
3  -0.012000000104308128   0.006000000052154064   0.01600000075995922   
4  -0.014999999664723873   0.017000000923871994  0.027000000700354576   

                    5407 5409  
0  -0.007000000216066837  NaN  
1   0.004999999888241291  NaN  
2                      ?  NaN  
3  -0.017999999225139618  NaN  
4   -0.04899999871850014  NaN  

[5 rows x 5410 columns]

In [38]:
data.describe()

Unnamed: 0  5409
count  31020.000000   0.0
mean   15736.369278   NaN
std     9067.996328   NaN
min        0.000000   NaN
25%     7899.750000   NaN
50%    15744.500000   NaN
75%    23588.250000   NaN
max    31419.000000   NaN

In [39]:
data.to_parquet(
    "~/task-2b-failures-in-cellular-apoptosis/data/df.parquet.gzip", compression="gzip"
)

In [4]:
data = pd.read_parquet("~/task-2b-failures-in-cellular-apoptosis/data/df.parquet.gzip")

In [41]:
data.head()

Unnamed: 0                     0                       1  \
0           0  -0.16099999845027924   -0.014000000432133675   
1           1  -0.15800000727176666  -0.0020000000949949026   
2           2                     ?                       ?   
3           3  -0.16899999976158142    -0.02500000037252903   
4           4  -0.18299999833106995   -0.050999999046325684   

                       2                      3                      4  \
0  0.0020000000949949026  -0.035999998450279236  -0.032999999821186066   
1  -0.012000000104308128   -0.02500000037252903  -0.012000000104308128   
2                      ?                      ?                      ?   
3  -0.009999999776482582   -0.04100000113248825   -0.04500000178813934   
4  -0.023000000044703484   -0.07699999958276749   -0.09200000017881393   

                       5                     6                     7  \
0   -0.09300000220537186   0.02500000037252903  0.004999999888241291   
1   -0.10599999874830246  0.013000000268220901  0.004999999888241291   
2                      ?                     ?                     ?   
3    -0.0689999982714653   0.03799999877810478  0.014000000432133675   
4  -0.014999999664723873   0.07100000232458115  0.027000000700354576   

                      8  ...                   5399                  5400  \
0                   0.0  ...   0.006000000052154064  0.013000000268220901   
1                   0.0  ...  0.0020000000949949026  -0.00800000037997961   
2                     ?  ...                      ?                     ?   
3   0.00800000037997961  ...    0.01899999938905239  0.009999999776482582   
4  0.019999999552965164  ...   0.050999999046325684  0.012000000104308128   

                   5401                  5402                  5403  \
0  0.020999999716877937  0.019999999552965164   0.01600000075995922   
1  0.007000000216066837  0.014999999664723873  -0.00800000037997961   
2                     ?                     ?                     ?   
3   0.02500000037252903   0.02500000037252903  0.020999999716877937   
4   0.05000000074505806   0.03799999877810478  0.050999999046325684   

                    5404                   5405                  5406  \
0  -0.010999999940395355   0.003000000026077032  0.009999999776482582   
1  -0.010999999940395355  -0.004000000189989805  0.013000000268220901   
2                      ?                      ?                     ?   
3  -0.012000000104308128   0.006000000052154064   0.01600000075995922   
4  -0.014999999664723873   0.017000000923871994  0.027000000700354576   

                    5407 5409  
0  -0.007000000216066837  NaN  
1   0.004999999888241291  NaN  
2                      ?  NaN  
3  -0.017999999225139618  NaN  
4   -0.04899999871850014  NaN  

[5 rows x 5410 columns]

The column Unnamed: 0 seems to be the index column. Most other columns seem to be floats. The column 5409 seems to contain many NaNs. There is one column that contains only question marks (as far as it can be seen here), maybe columns like this should be removed if it also contains question marks for all other columns that are not displayed here.

In [42]:
data.dtypes

Unnamed: 0      int64
0              object
1              object
2              object
3              object
               ...   
5404           object
5405           object
5406           object
5407           object
5409          float64
Length: 5410, dtype: object

In [43]:
data.select_dtypes(include="float64")

5409
0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
...     ...
31015   NaN
31016   NaN
31017   NaN
31018   NaN
31019   NaN

[31020 rows x 1 columns]

In [44]:
data.select_dtypes(include="object")

0                       1                       2  \
0        -0.16099999845027924   -0.014000000432133675   0.0020000000949949026   
1        -0.15800000727176666  -0.0020000000949949026   -0.012000000104308128   
2                           ?                       ?                       ?   
3        -0.16899999976158142    -0.02500000037252903   -0.009999999776482582   
4        -0.18299999833106995   -0.050999999046325684   -0.023000000044703484   
...                       ...                     ...                     ...   
31015   -0.004999999888241291   -0.023000000044703484    0.017999999225139618   
31016   0.0010000000474974513     0.00800000037997961  -0.0010000000474974513   
31017   0.0020000000949949026    0.013000000268220901  -0.0020000000949949026   
31018  -0.0020000000949949026   -0.012000000104308128   0.0020000000949949026   
31019   -0.004999999888241291   -0.027000000700354576    0.023000000044703484   

                           3                      4                      5  \
0      -0.035999998450279236  -0.032999999821186066   -0.09300000220537186   
1       -0.02500000037252903  -0.012000000104308128   -0.10599999874830246   
2                          ?                      ?                      ?   
3       -0.04100000113248825   -0.04500000178813934    -0.0689999982714653   
4       -0.07699999958276749   -0.09200000017881393  -0.014999999664723873   
...                      ...                    ...                    ...   
31015   -0.01600000075995922   -0.04500000178813934    0.05299999937415123   
31016   0.012000000104308128   0.019999999552965164  -0.019999999552965164   
31017    0.01600000075995922    0.03099999949336052  -0.026000000536441803   
31018  -0.017999999225139618  -0.027000000700354576   0.035999998450279236   
31019   -0.01899999938905239   -0.05000000074505806   0.050999999046325684   

                           6                      7                      8  \
0        0.02500000037252903   0.004999999888241291                    0.0   
1       0.013000000268220901   0.004999999888241291                    0.0   
2                          ?                      ?                      ?   
3        0.03799999877810478   0.014000000432133675    0.00800000037997961   
4        0.07100000232458115   0.027000000700354576   0.019999999552965164   
...                      ...                    ...                    ...   
31015   0.020999999716877937   0.003000000026077032   0.004000000189989805   
31016   -0.00800000037997961  -0.004000000189989805  -0.003000000026077032   
31017  -0.010999999940395355  -0.004999999888241291  -0.004000000189989805   
31018   0.013000000268220901   0.004999999888241291   0.006000000052154064   
31019   0.023000000044703484   0.003000000026077032   0.006000000052154064   

                            9  ...                   5398  \
0       -0.014999999664723873  ...  -0.014999999664723873   
1      -0.0020000000949949026  ...  -0.007000000216066837   
2                           ?  ...                      ?   
3       -0.014000000432133675  ...   -0.01600000075995922   
4        -0.01899999938905239  ...   -0.02800000086426735   
...                       ...  ...                    ...   
31015   -0.014999999664723873  ...  -0.054999999701976776   
31016    0.003000000026077032  ...   -0.04800000041723251   
31017    0.004999999888241291  ...   -0.04500000178813934   
31018   -0.004000000189989805  ...   -0.05400000140070915   
31019   -0.017000000923871994  ...   -0.04899999871850014   

                        5399                   5400                   5401  \
0       0.006000000052154064   0.013000000268220901   0.020999999716877937   
1      0.0020000000949949026   -0.00800000037997961   0.007000000216066837   
2                          ?                      ?                      ?   
3        0.01899999938905239   0.009999999776482582    0.02500000037252903   
4       0.050999999046325684   0.012000000104308128    0.0

The id column is an integer, the last column 5409 is a float and all others are objects at the moment.

In [45]:
np.unique(data["5409"])

array([nan])

The last column contains only NaN values.

In [47]:
df_exl = data.select_dtypes(exclude="float64")
df_exl.isna().to_numpy().any()

False

The columns apart from the 5409 column do not contain NaN values.

Next, I want to look into the question marks:

In [48]:
entry = data.iloc[2]
# pick all apart from first and last column that are not ?
selected = entry.to_numpy()[1:-1]
np.unique(selected)

array(['?'], dtype=object)

In [17]:
incomplete_entries = data.loc[data["2"] == "?"]

Found 261 rows with question marks. Not all of them have question marks in all columns. 

In [12]:
indices = []
for col in data.columns:
    idx = data.loc[data["2"] == "?"].index
    for id in idx:
        if id not in indices:
            indices.append(id)
print(indices)

[2, 15, 182, 184, 186, 187, 188, 196, 293, 392, 494, 690, 793, 990, 993, 1028, 1054, 1064, 1197, 1257, 1286, 1299, 1318, 1384, 1478, 1480, 1483, 1486, 1488, 1582, 1688, 1758, 1879, 1979, 2027, 2040, 2074, 2162, 2180, 2264, 2279, 2376, 2379, 2414, 2417, 2470, 2472, 2475, 2480, 2540, 2541, 2554, 2571, 2587, 2591, 2622, 2646, 2780, 2890, 2901, 3142, 3270, 3271, 3374, 3458, 3469, 3472, 3476, 3506, 3533, 3675, 3763, 3852, 3968, 4266, 4344, 4464, 4665, 4865, 4963, 5161, 5169, 5277, 5452, 5650, 6071, 6743, 6816, 6832, 7040, 7196, 7330, 7341, 7367, 7386, 7594, 7682, 7683, 7695, 7697, 7842, 7861, 7961, 7980, 8150, 9087, 9090, 9181, 9195, 9200, 9207, 9220, 9224, 9228, 9240, 9244, 9269, 9274, 9277, 9282, 9295, 9538, 9541, 9550, 9790, 9874, 10027, 10028, 10111, 10156, 10160, 10187, 10287, 10302, 10321, 10340, 10472, 10536, 11741, 11747, 11950, 12816, 12835, 13006, 13479, 13698, 14065, 14066, 14644, 14961, 15035, 15116, 15117, 15118, 15119, 15120, 15121, 15122, 15123, 15124, 15125, 15126, 15127, 15

In [13]:
len(indices)

261

Also, when iterating through all columns, not more than 261 entries with question marks have been found.

In [35]:
incomplete_entries
idx_mixed = []
idx_plain = []
for index, row in incomplete_entries.iterrows():
    selected = row.to_numpy()[1:-1]
    unique_vals = np.unique(selected)
    if len(unique_vals) == 1:
        idx_plain.append(index)
    else:
        idx_mixed.append(index)

print(f"There are {len(idx_plain)} entries that just contain '?'.")
print(f"There are {len(idx_mixed)} entries that contain other values apart from '?'.")

There are 137 entries that just contain '?'.
There are 124 entries that contain other values apart from '?'.


The 137 entries with just '?' and no other values should be removed from the dataframe, because they convey no information. For the 124 entries that have values for at least some columns a way to deal with the missing values needs to be found.

In [36]:
data = data.drop(index=idx_plain)

In [48]:
data

Unnamed: 0                       0                       1  \
0               0    -0.16099999845027924   -0.014000000432133675   
1               1    -0.15800000727176666  -0.0020000000949949026   
3               3    -0.16899999976158142    -0.02500000037252903   
4               4    -0.18299999833106995   -0.050999999046325684   
5               6    -0.15000000596046448     0.01600000075995922   
...           ...                     ...                     ...   
31015       31415   -0.004999999888241291   -0.023000000044703484   
31016       31416   0.0010000000474974513     0.00800000037997961   
31017       31417   0.0020000000949949026    0.013000000268220901   
31018       31418  -0.0020000000949949026   -0.012000000104308128   
31019       31419   -0.004999999888241291   -0.027000000700354576   

                            2                      3                      4  \
0       0.0020000000949949026  -0.035999998450279236  -0.032999999821186066   
1       -0.012000000104308128   -0.02500000037252903  -0.012000000104308128   
3       -0.009999999776482582   -0.04100000113248825   -0.04500000178813934   
4       -0.023000000044703484   -0.07699999958276749   -0.09200000017881393   
5       -0.014000000432133675                    0.0    0.01600000075995922   
...                       ...                    ...                    ...   
31015    0.017999999225139618   -0.01600000075995922   -0.04500000178813934   
31016  -0.0010000000474974513   0.012000000104308128   0.019999999552965164   
31017  -0.0020000000949949026    0.01600000075995922    0.03099999949336052   
31018   0.0020000000949949026  -0.017999999225139618  -0.027000000700354576   
31019    0.023000000044703484   -0.01899999938905239   -0.05000000074505806   

                           5                      6                       7  \
0       -0.09300000220537186    0.02500000037252903    0.004999999888241291   
1       -0.10599999874830246   0.013000000268220901    0.004999999888241291   
3        -0.0689999982714653    0.03799999877810478    0.014000000432133675   
4      -0.014999999664723873    0.07100000232458115    0.027000000700354576   
5       -0.12300000339746475  -0.004000000189989805  -0.0020000000949949026   
...                      ...                    ...                     ...   
31015    0.05299999937415123   0.020999999716877937    0.003000000026077032   
31016  -0.019999999552965164   -0.00800000037997961   -0.004000000189989805   
31017  -0.026000000536441803  -0.010999999940395355   -0.004999999888241291   
31018   0.035999998450279236   0.013000000268220901    0.004999999888241291   
31019   0.050999999046325684   0.023000000044703484    0.003000000026077032   

                           8  ...                   5399  \
0                        0.0  ...   0.006000000052154064   
1                        0.0  ...  0.0020000000949949026   
3        0.00800000037997961  ...    0.01899999938905239   
4       0.019999999552965164  ...   0.050999999046325684   
5      -0.004999999888241291  ...                    0.0   
...                      ...  ...                    ...   
31015   0.004000000189989805  ...   0.006000000052154064   
31016  -0.003000000026077032  ...  -0.009999999776482582   
31017  -0.004000000189989805  ...  -0.012000000104308128   
31018   0.006000000052154064  ...   0.014999999664723873   
31019   0.006000000052154064  ...   0.012000000104308128   

                        5400                   5401                   5402  \
0       0.013000000268220901   0.020999999716877937   0.019999999552965164   
1       -0.00800000037997961   0.007000000216066837   0.014999999664723873   
3       0.009999999776482582    0.02500000037252903    0.02500000037252903   
4       0.012000000104308128    0.05000000074505806    0.03799999877810478   
5      -0.032999999821186066   -0.03200000151991844   0.028999999165534973   
...                      ...                    ...                    ...   
31015    0.03

In [47]:
cols = data.columns
cols = cols.drop(labels="Unnamed: 0")
data[cols].drop_duplicates()

0                       1                       2  \
0        -0.16099999845027924   -0.014000000432133675   0.0020000000949949026   
1        -0.15800000727176666  -0.0020000000949949026   -0.012000000104308128   
3        -0.16899999976158142    -0.02500000037252903   -0.009999999776482582   
4        -0.18299999833106995   -0.050999999046325684   -0.023000000044703484   
5        -0.15000000596046448     0.01600000075995922   -0.014000000432133675   
...                       ...                     ...                     ...   
31015   -0.004999999888241291   -0.023000000044703484    0.017999999225139618   
31016   0.0010000000474974513     0.00800000037997961  -0.0010000000474974513   
31017   0.0020000000949949026    0.013000000268220901  -0.0020000000949949026   
31018  -0.0020000000949949026   -0.012000000104308128   0.0020000000949949026   
31019   -0.004999999888241291   -0.027000000700354576    0.023000000044703484   

                           3                      4                      5  \
0      -0.035999998450279236  -0.032999999821186066   -0.09300000220537186   
1       -0.02500000037252903  -0.012000000104308128   -0.10599999874830246   
3       -0.04100000113248825   -0.04500000178813934    -0.0689999982714653   
4       -0.07699999958276749   -0.09200000017881393  -0.014999999664723873   
5                        0.0    0.01600000075995922   -0.12300000339746475   
...                      ...                    ...                    ...   
31015   -0.01600000075995922   -0.04500000178813934    0.05299999937415123   
31016   0.012000000104308128   0.019999999552965164  -0.019999999552965164   
31017    0.01600000075995922    0.03099999949336052  -0.026000000536441803   
31018  -0.017999999225139618  -0.027000000700354576   0.035999998450279236   
31019   -0.01899999938905239   -0.05000000074505806   0.050999999046325684   

                           6                       7                      8  \
0        0.02500000037252903    0.004999999888241291                    0.0   
1       0.013000000268220901    0.004999999888241291                    0.0   
3        0.03799999877810478    0.014000000432133675    0.00800000037997961   
4        0.07100000232458115    0.027000000700354576   0.019999999552965164   
5      -0.004000000189989805  -0.0020000000949949026  -0.004999999888241291   
...                      ...                     ...                    ...   
31015   0.020999999716877937    0.003000000026077032   0.004000000189989805   
31016   -0.00800000037997961   -0.004000000189989805  -0.003000000026077032   
31017  -0.010999999940395355   -0.004999999888241291  -0.004000000189989805   
31018   0.013000000268220901    0.004999999888241291   0.006000000052154064   
31019   0.023000000044703484    0.003000000026077032   0.006000000052154064   

                            9  ...                   5399  \
0       -0.014999999664723873  ...   0.006000000052154064   
1      -0.0020000000949949026  ...  0.0020000000949949026   
3       -0.014000000432133675  ...    0.01899999938905239   
4        -0.01899999938905239  ...   0.050999999046325684   
5        0.009999999776482582  ...                    0.0   
...                       ...  ...                    ...   
31015   -0.014999999664723873  ...   0.006000000052154064   
31016    0.003000000026077032  ...  -0.009999999776482582   
31017    0.004999999888241291  ...  -0.012000000104308128   
31018   -0.004000000189989805  ...   0.014999999664723873   
31019   -0.017000000923871994  ...   0.012000000104308128   

                        5400                   5401                   5402  \
0       0.013000000268220901   0.020999999716877937   0.019999999552965164   
1       -0.00800000037997961   0.007000000216066837   0.014999999664723873   
3       0.009999999776482582    0.02500000037252903    0.02500000037252903   
4       0.012000000104308128    0.05000000074505806    0.03799999877810478   
5      -0.032999999821186066   -0.032000001519

When comparing all rows execpt for the index row unnamed 0, 5 entries get dropped as duplicated.

### External Info on Data

Link to dataset: https://archive.ics.uci.edu/dataset/188/p53+mutants
There are a total of 5409 attributes per instance.  
Attributes 1-4826 represent 2D electrostatic and surface based features.
Attributes 4827-5408 represent 3D distance based features.
Attribute 5409 is the class attribute, which is either active or inactive.